oak-core/src/main/java/org/apache/jackrabbit/oak/security/user/AuthorizablePropertiesImpl.java

```diff
@@ -234,9 +236,34 @@ private PropertyState getAuthorizableProperty(@NotNull Tree authorizableTree, @N
         }
         ReadOnlyNodeTypeManager nodeTypeManager = authorizable.getUserManager().getNodeTypeManager();
         PropertyDefinition def = nodeTypeManager.getDefinition(parent, property, true);
-        if (def.isProtected() || (authorizablePath.equals(parent.getPath())
-                && !def.getDeclaringNodeType().isNodeType(UserConstants.NT_REP_AUTHORIZABLE))) {
+        if (def.isProtected()) {
+            // exclude all protected properties
             return null;
+        } else if (authorizablePath.equals(parent.getPath())) {
+            // non-protected properties on the root must be defined by the expected
+            //  primary type or one of the configured mixin types
+            Boolean allowed = null;
+            NodeType declaringNodeType = def.getDeclaringNodeType();
+            if (declaringNodeType.isNodeType(UserConstants.NT_REP_AUTHORIZABLE)) {
+                // defined by the expected primary type so allowed
+                allowed = Boolean.TRUE;
```

Comment:
my understanding of the original improvement request is that allowing all types of properties is too permissive the the intention is to limit what properties can be stored below a given user/group node.

see the following sentence from the original report:
> This is in support of a use case where we want a stricter constraint on what is allowed to be stored in an authorizable property. 

given that `rep:Authorizables `allows for any residiual property of any type, single and multi-valued, this seems to defeat the purpose. my understanding of the request is that _only_ properties defined by the configured mixin types should be allow instead of allow any residual property as it would still be the case here.

as far as i can see this requirement isn't covered by the tests either.

Reply:
Yes, perhaps I didn't explain that well enough.  This proposal is mostly about reading authorizable properties.  I was able to deal with the setting of those authorizable properties using ACLs and already existing (and custom) code.

Basically my intention is to utilize ACLs to limit the setting of properties.  Essentially, I have re-configured the DefaultAuthorizableActionProvider (as below) so each user is only allowed "rep:alterProperties" rights on their home folder.  So they would only be able to update properties that already exist.  Then I have a custom AuthorizableActionProvider component that applies additional actions that will add the appropriate mixin type to the home folders and populates initial values for the properties that the end user should have access to alter.

For example, via OSGi configuration like this:
`        "org.apache.jackrabbit.oak.spi.security.user.action.DefaultAuthorizableActionProvider":{
            "userPrivilegeNames":[
                "jcr:read", 
                "rep:alterProperties"
            ],
            "groupPrivilegeNames":[
                "jcr:read"
            ],
            "enabledActions":[
                "org.apache.jackrabbit.oak.spi.security.user.action.AccessControlAction"
            ]
        }
`

Do you see any flaw with that approach?

Reply:
@enapps-enorman , let me try to rephrase to make sure i get your use case right:
- you want to make sure that a user can only modify and read the properties stored in in the subtree of it's home node
- upon user/group creation you create a set of properties
- but in addition you enforce particular types/values of those properties by adding mixins to the user/group node
- however, properties defined by those mixins will not be returned by Authorizable.getProperty.

so, your use-case is not about limiting the set of properties that can be written by adding mixins but rather making sure those defined by mixins are also returned as authorizable properties.
is that correct?

Reply:
@anchela Yes, that is basically correct.  My proposal is not really about limiting the set of properties that can be written, I simply wish to have a more specific value constraint defined for certain property definitions (as defined by a mixin type that I have assigned).  And I don't want the existence of that more specific property definition (with the value constraint) to make the property invisible from the Authorizable.getProperty call.

If you are curious, the more specific value constraint idea is motivated by security concerns about ensuring that self-registered end users of the platform can not abuse their self-managed user profile properties to store invalid or unreasonably long values to the repository.  For an extreme example, we don't want the user to do something like encoding a binary file as text and storing that value as the "nickname" property in their user profile.  We want some reasonable max length (or other) value constraint on the property definition so extremely invalid values will never be stored even if the end user somehow manages to get around the form post validation checking.

Reply:
@enapps-enorman , thanks for the insight.... and yes.... i am always curious to understand the reason behind a feature request or a patch. so thanks..... and it helps.

i am still wondering if it would make sense to automatically add the configured mixin types to the node that holds the authorizable-properties. that would avoid that developers have to make some assumptions about the way user properties are being stored in the repository.... in particular if the properties are stored with an relative path.... because i guess you would not want such huge binaries to be uploaded to <user-node>/some/subtree either, would you?

Reply:
Yes, it is true that we don't want invalid values in the subtree either.  However, I believe that scenario could already be handled with the previously existing code via ACLs (by denying the rights to create children) and/or via pre-creating the possible subtree nodes with specific primary types that defines what properties are allowed.  Only the properties of the root home folder could not be handled that way which is why I proposed this enhancement.

In fact the above is exactly what we are trying to do with a custom AuthorizableActionProvider component that pre-creates a "private" subfolder with a custom primary type (and an appropriate ACL) whenever a new user is created.